File: bend2/effs/tcp_connect.c
// TCP
// ===

// The socket is non-blocking for life. The connect's outcome is w->code:
// EINPROGRESS parked the computation until the socket was writable, and
// then SO_ERROR says how it ended. TCP.connect_within's park also has a
// deadline, and a wake that finds the socket still not writable came from
// the clock: the connect is still pending, so it fails with ETIMEDOUT, the
// kernel's own timeout error. A connect that has ended by the wake answers
// how it ended, even past the deadline.
static Term tcp_connect_more(Env e, IoWork* w) {
  int           fd  = (int)w->made;
  int           err = (int)w->code;
  socklen_t     len = sizeof(err);
  struct pollfd out = { fd, POLLOUT, 0 };
  if (err == EINPROGRESS && io_wait_time(w) != 0 && poll(&out, 1, 0) < 1) {
    err = ETIMEDOUT;
  }
  if (err == EINPROGRESS && getsockopt(fd, SOL_SOCKET, SO_ERROR, &err, &len)) {
    err = errno;
  }
  if (err != 0 && fd >= 0) {
    close(fd);
  }
  free(w->data);
  return err != 0 ? io_fail(e, (u32)err, NULL) : io_done(e, io_hand(fd));
}

// until is the park's deadline, an io_tick() (0: none).
static Term tcp_connect_start(Env e, Term* f, IoWork* w, u64 until) {
  struct sockaddr_in at;
  w->data = io_cstr(e, f[0], &w->size);
  int fd  = -1;
  errno   = EINVAL;
  if (!io_nul(w->data, w->size) && io_sys_addr(w->data, (u32)f[1], &at) == 0) {
    fd = socket(AF_INET, SOCK_STREAM, 0);
  }
  if (fd >= 0 && fcntl(fd, F_SETFL, fcntl(fd, F_GETFL) | O_NONBLOCK) < 0) {
    close(fd);
    fd = -1;
  }
  w->made = fd;
  io_sys_end(w, fd < 0 ? fd : connect(fd, (struct sockaddr*)&at, sizeof(at)));
  return w->code == EINPROGRESS ? io_wait_on(w, fd, POLLOUT, until,
    tcp_connect_more) : tcp_connect_more(e, w);
}

#ifdef CID_TCP_CONNECT

Term tcp_connect_run(Env e, Term* f, IoWork* w) {
  return tcp_connect_start(e, f, w, 0);
}

static void __attribute__((constructor)) tcp_connect_use(void) {
  io_eff(CID_TCP_CONNECT, tcp_connect_run, 0);
}

#endif

#ifdef CID_TCP_CONNECT_WITHIN

// TCP.connect with a deadline ms from now, as TCP.poll's.
Term tcp_connect_within_run(Env e, Term* f, IoWork* w) {
  return tcp_connect_start(e, f, w, io_tick() + (u64)f[2] * 1000000ull);
}

static void __attribute__((constructor)) tcp_connect_within_use(void) {
  io_eff(CID_TCP_CONNECT_WITHIN, tcp_connect_within_run, 0);
}

#endif

File: bend2/effs/tcp_connect.js
// TCP
// ===

// The socket is non-blocking for life: EINPROGRESS parks the computation
// until the socket is writable, and then SO_ERROR says how the connect ended.
// TCP.connect_within's park also has a deadline (until): a wake that finds
// the socket still not writable came from the clock, so the connect fails
// with ETIMEDOUT, and one that has ended by the wake answers how it ended.
function tcp_connect_with(host, port, until, k) {
  const sys = io_sys();
  const at = io_addr(host, Number(port));
  if (at === null) {
    return io_fail(22);
  }
  const fd = sys.socket(2, 1, 0);
  if (fd < 0) {
    return io_fail(sys.errno());
  }
  const end = (code) => {
    if (code !== 0) {
      sys.close(fd);
      return io_fail(code);
    }
    return io_done(fd);
  };
  const error = () => {
    const v = new Int32Array([0]);
    const l = new Uint32Array([4]);
    return sys.getsockopt(fd, sys.mac ? 0xffff : 1, sys.mac ? 0x1007 : 4,
      sys.ptr(v), sys.ptr(l)) < 0 ? sys.errno() : v[0];
  };
  // a select on the one fd that does not wait, its set sized as io_wait's
  const ready = () => {
    const out = new Uint8Array((fd >> 6 << 3) + 8);
    const tv = new BigInt64Array(2);
    out[fd >> 3] = 1 << (fd & 7);
    return sys.select(fd + 1, null, sys.ptr(out), null, sys.ptr(tv)) > 0;
  };
  const set = sys.fcntl(fd, 4, sys.fcntl(fd, 3, 0) | (sys.mac ? 4 : 0x800));
  const ok = set >= 0 && sys.connect(fd, sys.ptr(at), 16) >= 0;
  const code = ok ? 0 : sys.errno();
  if (code !== (sys.mac ? 36 : 115)) {
    return end(code);
  }
  io_park_on(fd, true, k, () => end(until === undefined || ready() ? error()
    : sys.mac ? 60 : 110), until);
  return undefined;
}

function tcp_connect(host, port, k) {
  return tcp_connect_with(host, port, undefined, k);
}

// TCP.connect with a deadline ms from now, as TCP.poll's.
function tcp_connect_within(host, port, ms, k) {
  return tcp_connect_with(host, port, performance.now() + Number(ms), k);
}
